jupyterhub/app.py

```diff
@@ -2303,6 +2334,20 @@ async def init_role_assignment(self):
         admin_role_objects = ['users', 'services']
         config_admin_users = set(self.authenticator.admin_users)
         db = self.db
+        # remove stale role assignments from authenticator
+        if self.authenticator.reset_managed_roles_on_startup:
+            for kind in kinds:
+                entity_name = kind[:-1]
+                association_class = orm.role_associations[entity_name]
+                deleted = (
+                    db.query(association_class)
+                    .filter(association_class.managed_by_auth == True)
+                    .delete()
+                )
+                if deleted:
+                    self.log.info(
+                        f"Deleted {deleted} stale {entity_name} role assignments previously added by an authenticator"
```

Comment:
This is a _highly_ disruptive operation - all users lose all permissions and are forced to login again to regain their access.

If this is the way we go, this flag should have a lot of caveats in its help that it should only be True only once in a while to reset permissions and usually switch back to False. Either that, or it should only be combined with a comprehensive `load_managed_roles`, and not be used otherwise.

Reply:
Right, especially because earlier we defined that `load_managed_roles` does not do role assignments, only role definitions. I will think about it more.

Reply:
> Right, especially because earlier we defined that load_managed_roles does not do role assignments, only role definitions. I will think about it more.

If it makes things simpler, I think it's okay to allow load_managed_roles to do assignments.

Reply:
This was addressed. I think the code is ready for another round of review.

> Either that, or it should only be combined with a comprehensive `load_managed_roles`, and not be used otherwise.

I now lean towards having `load_managed_roles` as always comprehensive.

> If it makes things simpler, I think it's okay to allow load_managed_roles to do assignments.

I was wrong, there was no such restriction.